### PR TITLE
Improved Linear Control Ops, with Numba Dispatches

### PR DESCRIPTION
**Pull request by jessegrabowski**
## Description
Another day, another PR with a huge pile of matrix decompositions nobody asked for :)

`solve_discrete_lyapunov` is quite important for statespace, because it's used to initialize a (big) covariance matrix. There are two modes for `solve_discrete_lyapunov`: direct and bilinear. We have bilinear support in the C backend, but Numba and JAX only support direct. Direct involves making a kronecker product of a (big) covariance matrix, so it sucks.

Several linear control Ops: `solve_sylvester`, `solve_discrete_lyapunov`, and `solve_continuous_lyapunov`, are all actually the same Op. They involve a Schur decomposition of the inputs, followed by a call to the TRSYL routine. Aside from that, it's just some dots and solves. 

This PR adds the `Schur` and `TRSYL` Ops, then uses these two components to implement all of the linear control ops (except for `solve_discrete_are`, that's another kettle of fish). These two new Ops don't have gradients, but the solvers that use them do, and `lop_override` is used to handle this. TRSYL is also not added to `__all__`, as it is not intended to be user-facing. `Schur` is a well-known matrix decomposition, so I exposed that one.

This has several advantages:

- By exposing the inner graphs (that were previously in scipy.linalg) we can do linag rewrites 
- Less code to maintain -- all of the linear control Ops now flow through `solve_sylvester`, so I removed all of the specialized Ops/gradients for the others
-  Free backend dispatching -- as long as we can implement Schur and TRSYL in a backend, we get all the rest of the stuff "for free".

As it turns out, we can implement Schur in Numba and JAX, and TRSYL in Numba. This PR also does that.

I think @Dekermanjian added some code to JAX that would let us dispatch these ops to jax as well, but I don't know if it's merged/released. I would be willing to do that in this PR if it's out.

## Related Issue

- [ ] Closes #
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

File: pytensor/link/numba/dispatch/linalg/decomposition/schur.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        N = val_to_int_ptr(_N)
+        LDA = val_to_int_ptr(_N)
+        SDIM = val_to_int_ptr(_N)
+        W = np.empty(_N, dtype=dtype)
+        _LDVS = _N
+        LDVS = val_to_int_ptr(_N)
+        VS = np.empty((_LDVS, _N), dtype=dtype)
+        RWORK = np.empty(_N, dtype=w_type)
+        BWORK = val_to_int_ptr(1)
+        INFO = val_to_int_ptr(1)
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/linalg/decomposition/schur.py:180]: Info isn't used?

File: pytensor/link/jax/dispatch/slinalg.py [hunk 2]
```diff
@@ -183,3 +184,17 @@ def expm(x):
...
 
     return expm
+
+
+@jax_funcify.register(Schur)
+def jax_funcify_Schur(op, **kwargs):
+    output = op.output
+
+    if op.sort is not None:
+        raise NotImplementedError("jax.scipy.linalg.schur only supports sort=None.")
```
Comment (ricardoV94) [inline pytensor/link/jax/dispatch/slinalg.py]: Instead of failing, issue a warning that it is ignored?

File: pytensor/link/numba/dispatch/linalg/solve/linear_control.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            LDA,
+            B_copy.ctypes,
+            LDB,
+            C_copy.ctypes,
+            LDC,
+            SCALE.ctypes,
+            INFO,
+        )
+
+        if int_ptr_to_val(INFO) < 0:
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/linalg/solve/linear_control.py:93]: `> 0` never happens or it's fine?

File: pytensor/link/numba/dispatch/linalg/solve/linear_control.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            return np.full_like(C_copy, np.nan)
+
+        # CC now contains the solution, scale it
+        X = SCALE * C_copy
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/linalg/solve/linear_control.py]: Reuse the array and then return it?

```suggestion
        C_copy *= SCALE
```

File: pytensor/link/numba/dispatch/slinalg.py [hunk 5]
```diff
@@ -469,3 +478,89 @@ def qr(a):
...
+
+
+@register_funcify_default_op_cache_key(Schur)
+def numba_funcify_Schur(op, node, **kwargs):
+    output = op.output
+    overwrite_a = op.overwrite_a
+    sort = op.sort
+
+    if sort is not None:
+        raise NotImplementedError(
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/slinalg.py]: Same, warn, or fallback to obj mode, don't fail

File: pytensor/tensor/slinalg.py [hunk 6]
```diff
@@ -1296,151 +1297,187 @@ def infer_shape(self, fgraph, node, shapes):
...
-        out_dtype = pytensor.scalar.upcast(A.dtype, B.dtype)
-        X = pytensor.tensor.matrix(dtype=out_dtype)
+        out_dtype = pytensor.scalar.upcast(A.dtype, B.dtype, C.dtype)
 
-        return pytensor.graph.basic.Apply(self, [A, B], [X])
+        output_shape = C.type.shape
+        if output_shape[0] is None and A.type.shape[0] is not None:
+            output_shape[0] = A.type.shape[0]
+        if output_shape[1] is None and B.type.shape[0] is not None:
+            output_shape[1] = B.type.shape[0]
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py]: bot is right

File: pytensor/tensor/slinalg.py [hunk 7]
```diff
@@ -1296,151 +1297,187 @@ def infer_shape(self, fgraph, node, shapes):
...
 
-    def perform(self, node, inputs, output_storage):
-        (A, B) = inputs
-        X = output_storage[0]
+        X = tensor(dtype=out_dtype, shape=output_shape)
 
-        out_dtype = node.outputs[0].type.dtype
-        X[0] = scipy_linalg.solve_continuous_lyapunov(A, B).astype(out_dtype)
+        return pytensor.graph.basic.Apply(self, [A, B, C], [X])
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py]: Import Apply (probably already imported) and use that directly

File: pytensor/tensor/slinalg.py [hunk 8]
```diff
@@ -1296,151 +1297,187 @@ def infer_shape(self, fgraph, node, shapes):
...
 
-        A_bar = S.dot(X.conj().T) + S.conj().T.dot(X)
-        Q_bar = -S  # Eq 29, adjusted
+        Y, scale, info = trsyl(A, B, C, overwrite_c=self.overwrite_c)
 
-        return [A_bar, Q_bar]
+        if info < 0:
+            return np.full_like(C, np.nan, dtype=out_dtype)
 
+        X[0] = scale * Y
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py]: Again do the mul inplace

File: pytensor/tensor/slinalg.py [hunk 9]
```diff
@@ -1296,151 +1297,187 @@ def infer_shape(self, fgraph, node, shapes):
...
-        Q_bar = S
-        return [A_bar, Q_bar]
+    op = SolveSylvester(
+        inputs=[A_matrix, B_matrix, Q_matrix],
+        outputs=[X],
+        lop_overrides=_lop_solve_continuous_slyvester,
+    )
 
+    return cast(
+        TensorVariable, Blockwise(op, signature="(m,m),(m,n),(n,n)->(m,n)")(A, B, Q)
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py]: You can put the gufunc_signature in the Op, then Blockwise will take it from there

File: pytensor/tensor/slinalg.py [hunk 10]
```diff
@@ -1477,12 +1514,29 @@ def solve_discrete_lyapunov(
...
 
     if method == "direct":
-        signature = BilinearSolveDiscreteLyapunov.gufunc_signature
-        X = pt.vectorize(_direct_solve_discrete_lyapunov, signature=signature)(A, Q)
-        return cast(TensorVariable, X)
+        vec_kron = pt.vectorize(kron, signature="(n,n),(n,n)->(m,m)")
+        AxA = vec_kron(A, A.conj())
+        eye = pt.eye(AxA.shape[-1])
+
+        vec_Q = Q.reshape((*Q.shape[:-2], -1))
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py]: `join_dims` *cof* *cof*

File: pytensor/tensor/slinalg.py [hunk 11]
```diff
@@ -2068,6 +2122,193 @@ def qr(
...
+    __props__ = ("output", "overwrite_a", "sort")
+
+    def __init__(
+        self,
+        output: Literal["real", "complex"] = "real",
+        overwrite_a: bool = False,
+        sort: Literal["lhp", "rhp", "iuc", "ouc"] | None = None,
+    ):
+        self.output = output
+        self.gufunc_signature = "(m,m)->(m,m),(m,m)"
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:1793]: if gufunc_signature is constant, define it as an Op attribute, instead of in `__init__`

File: tests/tensor/test_slinalg.py [hunk 12]
```diff
@@ -915,6 +917,68 @@ def test_expm_grad(mode):
     utt.verify_grad(expm, [A], rng=rng, abs_tol=1e-5, rel_tol=1e-5)
 
 
+@pytest.mark.parametrize(
+    "shape, use_complex",
+    [((5, 5), False), ((5, 5), True), ((5, 5, 5), False)],
+    ids=["float", "complex", "batch_float"],
+)
+def test_solve_continuous_sylvester(shape: tuple[int], use_complex: bool):
+    # batch-complex case got an error from BatchedDot not implemented for complex numbers
```
Comment (ricardoV94) [inline tests/tensor/test_slinalg.py:928]: then the rewrite that introduces BatchedDot shouldn't try to introduce it?

File: tests/tensor/test_slinalg.py [hunk 13]
```diff
@@ -915,6 +917,68 @@ def test_expm_grad(mode):
...
+    rng = np.random.default_rng()
+
+    dtype = config.floatX
+    if use_complex:
+        precision = int(dtype[-2:])  # 64 or 32
+        dtype = f"complex{int(2 * precision)}"
```
Comment (ricardoV94) [inline tests/tensor/test_slinalg.py]:
```suggestion
        dtype = "complex128" if dtype == "float64" else "complex64"
```

**Comment by Dekermanjian** [comment]
> I think @Dekermanjian added some code to JAX that would let us dispatch these ops to jax as well, but I don't know if it's merged/released. I would be willing to do that in this PR if it's out.

Hey @jessegrabowski, yes I did. I added the Sylvester solver. I believe you can implement the Lyapunov like so:

```
import jax
import jax.numpy as jnp
import jax.scipy as jsp

def solve_discrete_lyapunov(A, C, method="bilinear"):
    if method == "bilinear":
        eye = jnp.eye(A.shape[0])
        aH = A.conj().transpose()
        aHI_inv = jnp.linalg.inv(aH + eye)
        b = jnp.dot(aH - eye, aHI_inv)
        c = 2 * jnp.dot(jnp.dot(jnp.linalg.inv(A + eye), C), aHI_inv)
        solution = jsp.linalg.solve_sylvester(b.conj().transpose(), b.conj(), -c)
        return solution
    if method == "direct":
        lhs = jnp.kron(A, A.conj())
        lhs = jnp.eye(lhs.shape[0]) - lhs
        x = jsp.linalg.solve(lhs, C.flatten())
        return jnp.reshape(x, C.shape)
```

I would also be happy to open a PR and add this method directly to Jax.

**Comment by jessegrabowski** [comment]
> I would also be happy to open a PR and add this method directly to Jax.

If we have `jsp.linalg.solve_sylvester` I will just dispatch that, then we don't need a lyapunov dispatch at all. This PR removes the lyapunov Op, and essentially just does that code you posted as a pytensor graph (except using solve instead of inv), so we would be able to dispatch that whole graph, potentially with optimizations once we can infer matrix structures.

**Comment by Dekermanjian** [comment]
> > I would also be happy to open a PR and add this method directly to Jax.
> 
> If we have `jsp.linalg.solve_sylvester` I will just dispatch that, then we don't need a lyapunov dispatch at all. This PR removes the lyapunov Op, and essentially just does that code you posted as a pytensor graph (except using solve instead of inv), so we would be able to dispatch that whole graph, potentially with optimizations once we can infer matrix structures.

This is really super cool! Please let me know what I can do to help out!!

**Comment by jessegrabowski** [comment]
Drag that monster statespace PR over the finish line :P

File: pytensor/link/numba/dispatch/linalg/solve/linear_control.py (hunk 3 above)
Comment (jessegrabowski) [reply pytensor/link/numba/dispatch/linalg/solve/linear_control.py:93]: It can be 1, which means there were multiple eigenvalues of the same value, so a solution was found by jittering the matrix. A solution is returned in that case, so we can pass it back. For reference, scipy spams warnings in this case but also gives you back a matrix.

File: pytensor/tensor/slinalg.py [hunk 14]
```diff
@@ -1296,151 +1298,188 @@ def infer_shape(self, fgraph, node, shapes):
 expm = Blockwise(Expm())
 
 
-class SolveContinuousLyapunov(Op):
+class TRSYL(Op):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py]: There's a `tensor._linalg` module, don't be shy to start populating it instead of these `nlinalg` / `slinalg`

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py]: good idea!

File: tests/tensor/test_slinalg.py (hunk 12 above)
Comment (jessegrabowski) [reply tests/tensor/test_slinalg.py:928]: wdym?

**Comment by jessegrabowski** [comment]
Got everything passing. There was a strange failure for `link/numba/linalg/test_linear_control.py::test_solve_discrete_lyapunov`. The LU decomposition rewrite on general solve is introducing the `numba_pivot_to_permutation` function, which in turn errors with:

```
'numpy.dtypes.Int32DType' object has no attribute 'is_precise'
```

but only in object mode. I just disabled object mode, since it's not ever going to really be used. If anyone knows the real fix I'll do it, but I'm happy with where it's at.

**Comment by jessegrabowski** [comment]
Also this is should be tagged major because I moved imports for the linear control solvers (`tensor.slinalg.solve_discrete_lyapunov` etc) to just `tensor.linalg.solve_discrete_lyapunov`. I could put a small wrapper in slinalg with a deprecation warning? 

I kind of want a deprecation warning on all direct imports from `slinalg` and `nlinalg`, pending a refactor into a linalg sub-module. But that's neither here nor there.

**Comment by jessegrabowski** [comment]
Also for `jax.linalg.solve_sylvester` we need `jax>0.8`. We don't have a jax version pin, and the most recent version on conda-forge is only 0.7.2. I might bring back the rewrite for older jax versions? Or add a message to go update jax to use this Op, like the tfp error we use for those special functions?

File: pytensor/link/numba/dispatch/slinalg.py [hunk 15]
```diff
@@ -469,3 +478,88 @@ def qr(a):
...
     return qr, cache_version
+
+
+@register_funcify_default_op_cache_key(Schur)
+def numba_funcify_Schur(op, node, **kwargs):
+    output = op.output
+    overwrite_a = op.overwrite_a
+    sort = op.sort
+
+    if sort is not None:
```
Comment (ricardoV94) [inline pytensor/link/numba/dispatch/slinalg.py:489]: Maybe a compile verbose print of why it's going to fallback?

File: pytensor/tensor/_linalg/solve/linear_control.py [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+
+    def inplace_on_inputs(self, allowed_inplace_inputs: list[int]) -> "Op":
+        if not allowed_inplace_inputs:
+            return self
+        new_props = self._props_dict()  # type: ignore
+        new_props["overwrite_c"] = True
+        return type(self)(**new_props)
+
+
+def _trsyl(A: TensorLike, B: TensorLike, C: TensorLike) -> TensorVariable:
```
Comment (ricardoV94) [inline pytensor/tensor/_linalg/solve/linear_control.py]: This helper is useless?

File: tests/link/numba/linalg/solve/test_linear_control.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+from tests.link.numba.test_basic import compare_numba_and_py
+
+
+floatX = config.floatX
+
+pytestmark = pytest.mark.filterwarnings(
+    "ignore:numba.core.errors.NumbaPerformanceWarning"
+)
+
+numba = pytest.importorskip("numba")
```
Comment (ricardoV94) [inline tests/link/numba/linalg/solve/test_linear_control.py]: No longer need these as numba is now required dependency.

If needed I think it would need to come before the pytestmark

File: tests/tensor/test_slinalg.py (hunk 12 above)
Comment (ricardoV94) [reply tests/tensor/test_slinalg.py:928]: Replying to the comment in the first line of the test function

Comment (ricardoV94) [reply tests/tensor/test_slinalg.py:928]: Replying to the comment in the first line of the test function

Comment (jessegrabowski) [reply tests/tensor/test_slinalg.py:928]: ok i'll open an issue
